File: Editorial/0155_Min_Stack/CPP/2_Two_Stacks.cpp
class MinStack {
private:
    stack<int> stack, minStack;  // declaring the two stacks

public:
    MinStack() {}  // constructor

    void push(int x) {
        stack.push(x);  // pushing element to stack
        if (minStack.empty() || x <= minStack.top())
            minStack.push(x);  // pushing the minimum element so far to minStack
    }

    void pop() {
        if (stack.top() == minStack.top())
            minStack.pop();  // removing the top element of minStack if it's
                             // equal to the top of stack
        stack.pop();         // removing the top element of stack
    }

    int top() {
        return stack.top();  // returns the top element of stack
    }

    int getMin() {
        return minStack.top();  // returns the top element of minStack i.e.,
                                // minimum element up to now
    }
};

File: Editorial/0101-0200/0155_Min_Stack/JavaScript/2_Two_Stacks.js
function last(arr) {
    return arr[arr.length - 1];
}

class MinStack {
    _stack = [];
    _minStack = [];

    push(x) {
        this._stack.push(x);
        if (this._minStack.length === 0 || x <= last(this._minStack)) {
            this._minStack.push(x);
        }
    }

    pop() {
        if (last(this._minStack) === last(this._stack)) {
            this._minStack.pop();
        }
        this._stack.pop();
    }

    top() {
        return last(this._stack);
    }

    getMin() {
        return last(this._minStack);
    }
}
